### PR TITLE
Provide a way to explicitly invoke the debugger

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
Provide a way to explicitly invoke the debugger with in the debug strategy

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME

lib/ansible/plugins/strategy/debug.py
##### ANSIBLE VERSION

```
devel
```

##### ADDITIONAL INFORMATION

This change gives the user a way to explicitly invoke the debugger when using the debug strategy:
```
---
- hosts: localhost
  gather_facts: false
  tasks:
    - debug:
        msg: foo
      when: False
      debugger: true
```

We likely need some docs.

### CODE REVIEW

**Comment by sivel** [comment]
I think this is probably ready.  I've let it rest for a while, and have come back to it, and it seems complete at this point, just need some reviews.

File: lib/ansible/executor/stats.py [hunk 1]
```diff
@@ -46,6 +46,12 @@ def increment(self, what, host):
         prev = (getattr(self, what)).get(host, 0)
         getattr(self, what)[host] = prev + 1
 
+    def decrement(self, what, host):
+        try:
+            getattr(self, what)[host] -= 1
```
Comment (sivel) [inline lib/ansible/executor/stats.py]: Need to guard against this going negative

**Review by jimi-c — APPROVED** [review]
:+1: pending docs changes to discuss implications for the free strategy.

File: lib/ansible/plugins/strategy/__init__.py [hunk 2]
```diff
@@ -949,3 +1026,106 @@ def update_active_connections(self, results):
...
+    """ The next action after an interpreter's exit. """
+    REDO = 1
+    CONTINUE = 2
+    EXIT = 3
+
+    def __init__(self, result=EXIT):
+        self.result = result
+
+
+class Debugger(cmd.Cmd):
```
Comment (alikins) [inline lib/ansible/plugins/strategy/__init__.py:1041]: I wouldn't mind seeing the debugging implementation bits extract to a separate python file (lib/ansible/utils/strategy_debugger.py ?).  That is assuming adding python files to strategy/ that dont have a StrategyModule in them is a problem.

File: lib/ansible/plugins/strategy/__init__.py [hunk 3]
```diff
@@ -949,3 +1026,106 @@ def update_active_connections(self, results):
                 if socket_path:
                     if r._host not in self._active_connections:
                         self._active_connections[r._host] = socket_path
+
+
+class NextAction(object):
+    """ The next action after an interpreter's exit. """
+    REDO = 1
+    CONTINUE = 2
+    EXIT = 3
```
Comment (alikins) [inline lib/ansible/plugins/strategy/__init__.py:1035]: 👍 for enum

**Pull request merged**